README.md
# ubuntu-18.04-mate

Vagrant Box with Ubuntu 18.04 and mate

## Base image

Used base image [elegoev/ubuntu-18.04](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04)

## Directory Description

| directory | description                                          |
|-----------|------------------------------------------------------|
| inspec    | inspec test profiles with controls                   |
| packer    | packer build, provisioner and post-processor scripts |
| test      | test environment for provision & inspec development  |

## Configuration

### Vagrant Cloud

- [elegoev/ubuntu-18.04-mate](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04-mate)

### Useful Vagrant Plugins

- [vagrant-disksize](https://github.com/sprotheroe/vagrant-disksize)
- [vagrant-hosts](https://github.com/oscar-stack/vagrant-hosts)
- [vagrant-secret](https://github.com/tcnksm/vagrant-secret)
- [vagrant-vbguest](https://github.com/dotless-de/vagrant-vbguest)
- [vagrant-serverspec](https://github.com/vvchik/vagrant-serverspec)
- [vagrant-vmware-esxi](https://github.com/josenk/vagrant-vmware-esxi)

### Using mate

#### Vagrantfile

    Vagrant.configure("2") do |config|

      $basebox_name="ubuntu-18.04-mate-test"
      $basebox_hostname="ubuntu-1804-mate-test"
      $src_image_name="elegoev/ubuntu-18.04-mate"
      $vb_group_name="basebox-mate-test"

      config.vm.define "#{$basebox_name}" do |machine|
        machine.vm.box = "#{$src_image_name}"
    
        # define guest hostname
        machine.vm.hostname = "#{$basebox_hostname}"

        machine.vm.provider "virtualbox" do |vb|
          vb.name = $basebox_name
          vb.cpus = 1
          vb.customize ["modifyvm", :id, "--memory", "1024" ]
          vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
          vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
          vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
          vb.customize ["modifyvm", :id, "--vram", 256 ]
        end

        # forwarding ports
        machine.vm.network "forwarded_port", id: "rdp",  auto_correct: true, protocol: "tcp", guest: 3389, host: 33333, host_ip: "127.0.0.1"

      end   

    end

### Reference

- [How to set up an xRDP server on Ubuntu 18.04](https://medium.com/@vivekteega/how-to-setup-an-xrdp-server-on-ubuntu-18-04-89f7e205bd4e)

### Versioning

Repository follows sematic versioning  [![semantic versioning](https://img.shields.io/badge/semver-2.0.0-green.svg)](http://semver.org)

### Changelog

For all notable changes see [CHANGELOG](https://github.com/elegoev/basebox-ubuntu-18.04-mate/blob/master/CHANGELOG.md)

### License

Licensed under The MIT License (MIT) - for the full copyright and license information, please view the [LICENSE](https://github.com/elegoev/basebox-ubuntu-18.04-mate/blob/master/LICENSE) file.

### Issue Reporting

Any and all feedback is welcome.  Please let me know of any issues you may find in the bug tracker on github. You can find it [here.](https://github.com/elegoev/basebox-ubuntu-18.04-mate/issues)


CHANGELOG.md
# Change Log

## 1.0.x

**Implemented enhancements:**

- Mate installation

## 2.0.x

**Implemented enhancements:**

- Packer build implemented


boxtests/xrdp_spec.rb
# Test ansible
describe package('xrdp') do
    it { should be_installed }
end


provisioning/ubuntu-18.04-mate.sh
#!/bin/bash

# install apt-add-repository
sudo apt-get update
echo n | sudo DEBIAN_FRONTEND=noninteractive apt-get upgrade -y

# Install xrdp
sudo apt-get install -y xrdp
sudo cp /vagrant/files/apps/xrdp/km-00000807.ini /etc/xrdp/km-00000807.ini
sudo chmod 644 /etc/xrdp/km-00000807.ini

# set session
echo mate-session> /home/vagrant/.xsession
sudo chown vagrant /home/vagrant/.xsession
sudo chgrp vagrant /home/vagrant/.xsession

# restart xrdp
sudo service xrdp restart

# set keyboard input
# sudo loadkeys ch
L='ch' && sudo sed -i 's/XKBLAYOUT=\"\w*"/XKBLAYOUT=\"'$L'\"/g' /etc/default/keyboard

# create date string
DATE=`date +%Y%m%d%H%M`

# set version
UBUNTU_VERSION=$(lsb_release -a | grep Release | awk  '{print $2}')
echo "ubuntu-$UBUNTU_VERSION" > /vagrant/version


packer/provisioners/base-provisioning.sh
#!/bin/bash

application_file_path="/vagrant/installed-application.md"

# install mate
#sudo apt-get install -y mate-core mate-desktop-environment mate-notification-daemon
sudo apt install -y expect
cat <<EOF | expect
set timeout -1
spawn sudo apt install -y ubuntu-mate-desktop
expect "Default display manager: "
send "lightdm\n"
expect eof
EOF

# install language pack
sudo apt-get -y install `check-language-support -l de`

# set locale
sudo update-locale LANG=de_CH.UTF-8

# install xrdp
sudo apt install xrdp -y
sudo apt install xorg-video-abi-24 -y
sudo apt install xserver-xorg-input-all -y
sudo apt install xorgxrdp -y

# grant access
sudo adduser xrdp ssl-cert

# restart xrdp
sudo systemctl stop xrdp
sudo systemctl start xrdp

# set password for ubuntu user
echo -e "ubuntu\nubuntu" | sudo passwd ubuntu

# set .xsession
echo mate-session > /home/vagrant/.xsession
sudo chmod +x /home/vagrant/.xsession
sudo chown vagrant:vagrant /home/vagrant/.xsession

sudo cp /home/vagrant/.xsession /home/ubuntu/.xsession
sudo chmod +x /home/ubuntu/.xsession
sudo chown ubuntu:ubuntu /home/ubuntu/.xsession

# copy keymap
sudo cp /home/vagrant/files-prov/xrdp/km-00000807.ini /etc/xrdp/km-00000807.ini
sudo chmod 644 /etc/xrdp/km-00000807.ini

# copy polkit rule
sudo cp /home/vagrant/files-prov/xrdp/45-allow-colord.pkla /etc/polkit-1/localauthority/50-local.d/45-allow-colord.pkla

# set version
UBUNTU_VERSION=$(lsb_release -a | grep Release | awk  '{print $2}')
echo "# Installed application "  > $application_file_path
echo "***                     " >> $application_file_path
echo "> ubuntu: $UBUNTU_VERSION with mate and xrdp" >> $application_file_path


packer/post-processors/create-git-tag.ps1
# create git tag

$src_image_version = $env:IMAGE_VERSION

$vagrant_file_template = '
Vagrant.configure("2") do |config|

    $basebox_name="ubuntu-18.04-mate-test"
    $basebox_hostname="ubuntu-1804-mate-test"
    $src_image_name="elegoev/ubuntu-18.04-mate"
    $vb_group_name="basebox-mate-test"

    config.vm.define "#{$basebox_name}" do |machine|
      machine.vm.box = "#{$src_image_name}"
      machine.vm.box_version = "$image_version"
    
      # define guest hostname
      machine.vm.hostname = "#{$basebox_hostname}"

      machine.vm.provider "virtualbox" do |vb|
        vb.name = $basebox_name
        vb.cpus = 1
        vb.customize ["modifyvm", :id, "--memory", "4096" ]
        vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
        vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
        vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
        vb.customize ["modifyvm", :id, "--vram", 256 ]
      end
  
      # forwarding ports
      machine.vm.network "forwarded_port", id: "rdp",  auto_correct: true, protocol: "tcp", guest: 3389, host: 33333, host_ip: "127.0.0.1"

    end   

end
'

# create Vagrantfile for download test
$vagrant_file_expanded = $vagrant_file_template.Replace('$image_version', "$src_image_version")
Set-Content -Path './test/download/Vagrantfile' -Value "$vagrant_file_expanded"

# create git tag
git add --all
git commit -m "new basebox version $src_image_version"
git push 
git tag $src_image_version
git push --tags





